content/blog/2020-11-11-self/index.md

```diff
@@ -0,0 +1,155 @@
+
+They also use two other tricks to speed up heap scans. First, at the end of the heap region, they place a dummy "sentinel" reference that gets the attention of the scanner. This way, the scanner only needs to check if it's within bounds when it sees a reference instead of every time it advances itself by one word. Similarly, a scanner often needs to capture the object which references another object. To minimize bookkeeping, every object begins with a special tag called a "mark". Then, when a reference is found, the scanner can just iterate backward to the closest mark to find the referencing object.
+
+Segregation, sentinels, and marks improve from the heap scanning speed of the fastest Smalltalk implementation by a factor of two.
+
+### Object Formats
+
+32-bit words in the slot contents can be one of three things: 30-bit integers, 30-bit floats, or 32-bit addresses. The two least significant bits encode which type of data the word represents. As far as the memory address is concerned, this type of encoding is an example of a [tagged pointer][taggedptr]. The last two bits are also used to differentiate the "mark" words described in the previous section, although these words are never placed inside slots.
+
+This section of the paper goes on to show how objects and maps are layed out in memory. I don't think there's anything too interesting or clever about this, but the kicker is that at the end the authors can demonstrate concretely that by using maps, their Self implementation is as space efficient as Smalltalk.
+
+## The Parser
+
+Self uses a JIT, and therefore Self code is compiled into bytecode. Method objects store the bytecode as data using a byte array and an object array. Self uses a very simple stack-based bytecode representation with only eight opcodes (four of them being used to originate or delegate messages). In contrast, Smalltalk-80 has a much more eleborate bytecode representation.
+
+## The Compiler
+
+Like the state-of-the-art Smalltalk implementation of the day, the Self implementation uses *dynamic translation* (again, JIT), and *inline caching* of message results.
+
+Smalltalk programs are notoriously hard to optimize because the types of everything except primitives are not really known statically. Self has the same issue, compounded by the fact that variables are only accessible through message passing, which further obfuscates the compiler's view. In addition, Self cannot use Smalltalk optimizations which operate on classes. Nonetheless, the novel techniques described in this section make Self twice as efficient as the state-of-the-art Smalltalk implementation, or about five times slower than optimized C. I will point out that later implementations of Self acheived speeds only twice as slow as optimized C!
```

Comment:
Perhaps worth noting that this challenge exists in JavaScript, Python, Ruby, etc., i.e., every modern dynamic language. And a version of it exists in statically typed OO languages.